enclave_build/src/cache.rs

```diff
@@ -0,0 +1,764 @@
+use sha2::Digest;
+
+use crate::{
+    image::{self, ImageDetails},
+    EnclaveBuildError, Result,
+};
+
+/// Root folder for the cache.
+pub const CACHE_ROOT_FOLDER: &str = "XDG_DATA_HOME";
+/// Path to the blobs folder
+pub const CACHE_BLOBS_FOLDER: &str = "blobs/sha256/";
+/// Name of the cache index file which stores the (image URI <-> image hash) mappings.
+pub const CACHE_INDEX_FILE_NAME: &str = "index.json";
+/// The name of the OCI layout file from the cache.
+pub const CACHE_OCI_LAYOUT_FILE: &str = "oci-layout";
+
+/// Constants used for complying with the OCI cache structure
+pub const DEFAULT_MEDIA_TYPE: &str = "application/vnd.oci.image.manifest.v1+json";
+pub const REF_ANNOTATION: &str = "org.opencontainers.image.ref.name";
+pub const OCI_LAYOUT: (&str, &str) = ("imageLayoutVersion", "1.0.0");
+
+/// Manifest entry in index file. The digest will point to the actual manifest file
+#[derive(Debug, Serialize, Deserialize, PartialEq, Eq)]
+pub struct ManifestEntry {
+    #[serde(rename = "mediaType")]
+    media_type: String,
+    #[serde(rename = "size")]
+    size: usize,
+    #[serde(rename = "digest")]
+    digest: String,
+    #[serde(rename = "annotations")]
+    annotations: HashMap<String, String>,
+}
+
+/// Cache index file structure following OCI spec:
+/// https://github.com/opencontainers/image-spec/blob/main/image-layout.md#indexjson-file
+/// This file is used to search image manifests among blobs. Image names will be mapped
+/// to manifest configs
+#[derive(Debug, Serialize, Deserialize, PartialEq, Eq)]
+pub struct CacheIndex {
+    #[serde(rename = "schemaVersion", default = "schema_default")]
+    schema_version: u32,
+    #[serde(rename = "manifests", default = "Vec::new")]
+    manifests: Vec<ManifestEntry>,
+}
+
+impl Default for CacheIndex {
+    fn default() -> Self {
+        CacheIndex {
+            schema_version: 2,
+            manifests: Vec::new(),
+        }
+    }
+}
+
+fn schema_default() -> u32 {
+    2
+}
+
+/// Struct which provides operations with the local cache.
+///
+/// The index file is located in the cache root folder and keeps track of images stored in cache.
+/// The cache will be consumed during image operations (pull and inspect) to prevent requests to the
+/// image registry. Also `linuxkit` will use this cache as its own to prevent subsequent requests
+///
+/// The cache structure is:
+///
+/// {CACHE_ROOT_PATH}/index.json\
+/// {CACHE_ROOT_PATH}/blobs/sha256\
+/// {CACHE_ROOT_PATH}/blobs/sha256/hash_of_blob_1\
+/// {CACHE_ROOT_PATH}/blobs/sha256/hash_of_blob_2\
+/// etc.
+///
+/// A blob can be:
+///
+/// {IMAGE_FOLDER_PATH}/blobs/sha256/manifest_hash - the image manifest, stored as a JSON String.\
+/// {IMAGE_FOLDER_PATH}/blobs/sha256/config_hash - the image configuration, stored as a JSON String.\
+/// {IMAGE_FOLDER_PATH}/blobs/sha256/layer_hash - one of the image layers, each in a separate gzip compressed tar file.
+#[derive(Clone)]
+pub struct CacheManager {
```

Comment:
I feel we have a design problem here. For example we are doing the same operations multiple times, e.g. instead of fetching the cache index file multiple times, we should do that just once, and keep it in the cache manager (as a separate field).
I'll try to think of other ways of how we can improve the cache manager abstraction.